### PR TITLE
ujson to load ticker files 30% faster in BT

### PR DESCRIPTION
**Pull request by creslinux**
ujson loads ticker files 30% quicker in backest 
This can accumulate to many second as exchanges host 150+ pairs and users backtest longer data

In a test on 60 pairs with 2 weeks back data profile demonstrates load ticker time reduced from 
over 5 to 3.5 seconds. 

screen shot 2018-07-12 at 12 05 28 pm

screen shot 2018-07-12 at 12 04 09 pm

the change has been tech tested to show no impact to trades in backtest with stop/roi and buy/sell 1
no detla witnessed, diff linked

http://www.mergely.com/0qwuibQH/

### CODE REVIEW

File: freqtrade/optimize/__init__.py [hunk 1]
```diff
@@ -2,6 +2,7 @@
 
 import gzip
 import json
+import ujson
```
Comment (xmatthias) [inline freqtrade/optimize/__init__.py]: if you use `import ujson as json` it's a dropin-replacement (and should not break tests).

**Comment by xmatthias** [comment]
I suspect you could replace json where it's used, if we require this library anyway (`misc.py`, ...).

Not sure about the Failure - seems to be a precision "mismatch" ... maybe this handles data slightly different?

File: requirements.txt [hunk 2]
```diff
@@ -17,6 +17,7 @@ pytest-mock==1.10.0
 pytest-cov==2.5.1
 tabulate==0.8.2
 coinmarketcap==5.0.3
+ujson==1.35
```
Comment (xmatthias) [inline requirements.txt]: This increases the requirements on windows hugely ... 
```
error: Microsoft Visual C++ 14.0 is required. Get it with "Microsoft Visual C++ Build Tools": http://landinghub.visualstudio.com/visual-cpp-build-tools ... 
```
(link does not work anymore ...) - but it requires installing visual studio.

Not sure that's worth the speedup for something rarely used during day-to-day operations...

Comment (creslinux) [reply requirements.txt]: wow - yeh, totally not worth it. 
The failed test was on cached data, not needed for the loading speedup. 

Agree its not suitable if windows users need to install the world, how are they managing other compiled c, such as TA-Lib?

Comment (xmatthias) [reply requirements.txt]: ta-lib we have a instruction to install egg (it's not straightforward on linux either as you need the OS package ) ... 
didn't find one for this library though - but i'm at work, so didn't look too hard

**Comment by creslinux** [comment]
Made optional, out by default in req.txt
optimize tests for its presence uses if available.

File: requirements.txt (hunk 2 above)
Comment (creslinux) [reply requirements.txt]: are these them?
https://www.lfd.uci.edu/~gohlke/pythonlibs/#ujson

File: freqtrade/optimize/__init__.py [hunk 3]
```diff
@@ -12,6 +11,11 @@
 from freqtrade.exchange import Exchange
 from freqtrade.arguments import TimeRange
 
+import importlib
+ujson_found = importlib.util.find_spec("ujson")
+if ujson_found is not None:
+    import ujson
```
Comment (xmatthias) [inline freqtrade/optimize/__init__.py]: if you use `import ujson as json` it will act as a drop-in replacement ... so you don't need to check below

**Comment by creslinux** [comment]
Top top, thanks
Out walking the dog I'll do that when back

File: freqtrade/optimize/__init__.py (hunk 3 above)
Comment (xmatthias) [reply freqtrade/optimize/__init__.py]: + add a debug-logentry showing this was used ...

Comment (macd2) [reply freqtrade/optimize/__init__.py]: tested this option and works great just replaced:
`import json` with `import ujson as json`

**Comment by creslinux** [comment]
When using as a drop in to replace all json calls in optimise
The test "timeframe starts in the center of the cached data" fails

The failure is the precision is greater in ujson / rounded in json
Other tests are not affected. 

I'll have a look to see if precision can be passed to loads 

screen shot 2018-07-12 at 3 03 02 pm

**Comment by creslinux** [comment]
Old test data is on the right, the newly loaded data on the left has the greater precision. 
Only in the test when loading half way into a cached data test 

Its not as easy as setting precision, as the original test data is also not consistent in the number of places it displays.

**Comment by creslinux** [comment]
Data is aligned,
TL;DR - C world says python json is the mathematical equivalent of a banana, and begrudgingly allow a kw arg to be break math and align with pythons other json implementations  

[https://github.com/esnme/ultrajson/issues/69]

I'll now back test to see if the KW interferes with normal json or is gracefully ignored. 

`pairdata = json.load(tickerdata, precise_float=True)`

screen shot 2018-07-12 at 3 53 46 pm

**Comment by gcarq** [comment]
@creslinux maybe it helps to use `Decimal` for float parsing, but I'm not sure if this will break some other tests.
```
json.load(tickerdata, parse_float=decimal.Decimal)
```

File: freqtrade/optimize/__init__.py (hunk 3 above)
Comment (creslinux) [reply freqtrade/optimize/__init__.py]: @macd2 you may want to DL the latest commit. 
FP is handled differently to python json without a kwarg now added.

Comment (gcarq) [reply freqtrade/optimize/__init__.py]: As already mentioned by @xmatthias the check in `trim_tickerlist` should be removed and handled via imports, for example:
```
try:
    import ujson as json
except ImportError:
    import json
```

**Comment by creslinux** [comment]
@gcarq thanks, adding precise_float=true as a KW are to ujson.load aligns the two implementations. 

They have a long thread on it here https://github.com/esnme/ultrajson/issues/69 
The kwarg flips the C internally to use 'strod' for fp, which is what python uses. (i read)

**Comment by creslinux** [comment]
i was surprised when you said Visual studio needed under nt. On my mac is was just a pip3 install ujson - it was there, no compiling needed

**Comment by xmatthias** [comment]
it's not the "full" Visual studio, but part of it (~1.5Gb if i remember correctly when i checked for talib) ....

**Comment by vertti** [comment]
Hyperopt strategy loading is also broken under Windows currently. And because the WLS works well under Windows, I'm fine with dropping the regular windows support.

**Comment by gcarq** [comment]
I'm also fine with dropping windows support. As already said for Windows 10 WSL can be used and for Windows 7 cygwin or minGW

**Comment by xmatthias** [comment]
As the last comments show, we are happy to drop support for "native" windows - so i would remove all boilerplate, add `ujson` regularily to the requirements.txt and default to ujson.

Sorry for the confusion @creslinux

**Comment by creslinux** [comment]
Could use it only on load, our input format is fixed so no new bugs.

That's where we'll get the largest time saving too.

Or bite the bullet and move to feather or pickle

**Comment by xmatthias** [comment]
Will close this as #1090 solves the same issue without dropping Windows support and in a pythonic way ... :)
